### DISCUSSION TITLE
preact-compat

### DISCUSSION THREAD

**Discussion by jonasgroendahl** [freshframework/fresh#432] [Q&A]
How to use deno fresh with `preact-compat`? 

I'm trying to add zustand to a **_fresh_** project.

Normally you'd alias react and react-dom as explained here: https://preactjs.com/guide/v10/switching-to-preact but I don't quite get how to set that up for a fresh project.

**Comment by tylrw** [comment]
https://esm.sh URLs support an `alias` query parameter.

e.g. https://esm.sh/zustand?alias=react:preact/compat&deps=react

I'd still like to know what the best practice is.

**Reply by nikableh** [reply]
It works with [react-icons](https://www.npmjs.com/package/react-icons) and I am super hyped about this!

How does it work tho? Teach me.

**Reply by jonasgroendahl** [reply]
Teach me too about react icons 

Thanks for the help with Zustand!

**Reply by nikableh** [reply]
With react-icons, I got into [this](https://github.com/denoland/deno/issues/15108) trouble. And I also can't figure out how to extract `IconType` type. Using Node packages in Deno is still pretty painful...

**Reply by tylrw** [reply]
@lucacasonato might have insight on the right approach. I'm sure he's not busy 😄. 

<img width="738" alt="image" src="https://user-images.githubusercontent.com/7053824/177608065-66454ca3-1949-4dc4-8a48-3779939fcdd0.png">

**Comment by czottmann** [comment]
That's a pretty neat idea, @rtwall, thanks! I finally figured out how to tell a _dependency_ to use Preact instead of React, too!

I wanted to import [Statery](https://github.com/hmans/statery) ("surprise-free state management!") which internally uses React (excerpt from its [`index.ts`](https://github.com/hmans/statery/blob/master/src/index.ts#L1)):

```typescript
import { useEffect, useRef, useState } from "react"`)
```

So, in a React-based app I'd call it like this:

```typescript
import { makeStore, useStore } from "statery";
```

In Fresh, I instead use the esm.sh aliasing as described:

```typescript
import { makeStore, useStore } from "https://esm.sh/statery@0.5.4?alias=react:preact/compat&deps=preact@10.8.2";
```

If I omit the `deps=` argument the aliasing assumes I want the latest version of Preact (10.10.x), which clashes with the 10.8.2 used in Fresh (at the time of this writing).

So: aliasing works absolutely fine, and it's a good-enough practice until I'm told otherwise by people more knowledgeable than myself.  :)  I've learned something today, thanks.

**Reply by kidonng** [reply]
esm.sh recently added support for [external dependencies](https://esm.sh/#specify-external-dependencies), so instead of `&deps=preact@10.8.2` you can just use `&external=preact`.

**Reply by OldManFroggy** [reply]
you can also alias 'react' to preact compat, in your import_map.json this seems to work for a fair few react libs, that do not rely on features that are not in preact. (still a bit hit and miss).

**Reply by czottmann** [reply]
In this particular case …

```typescript
import { makeStore, useStore } from "https://esm.sh/statery@0.5.4?alias=react:preact/compat&deps=preact@10.8.2";
```

… using `&external=preact` (just replacing `&deps=…`) didn't replace the occurrence in the module loaded from esm.sh; but `&external=preact/compat` did the trick.

Alas, my code didn't run, though. :)  Had to go with my original URL. If you have more insights regarding `external=`, please let me know!

**Reply by czottmann** [reply]
@scott-ling I try to avoid overly implicit code like this — I tend to forget about the hacks that make my code work, and in a month I get confused.  ;)

**Reply by kidonng** [reply]
> Alas, my code didn't run, though. :) Had to go with my original URL.

Could be a recent bug: https://github.com/ije/esm.sh/issues/383

You can try old versions of esm.sh like `&pin=v87`.

**Comment by OldManFroggy** [comment]
completely agree, but right now a lot of tooling is not available to Fresh,
especially in the area of UI components, so these hacks are unfortunately
needed at least for now -> although I am looking at doing a fresh/preact
port of tailwind headless ui that does not rely on preact-compat
eventually.  (though I do wish Mantine worked with Preact/Fresh... oh
well...)

**Reply by dagda1** [reply]
This nugget of information `?alias=react:preact/compat&deps=react` is hidden away, I was just about to throw in the towel with fresh until I found this.  It should be more prominent somewhere

**Reply by andromidasj** [reply]
I've been racking my brain over getting Mantine to work with Preact & Fresh, too. Have you had any luck? I've tried importing it with the alias query param:
```
{
  "imports": {
    "$fresh/": "https://deno.land/x/fresh@1.1.2/",
    "preact": "https://esm.sh/preact@10.11.0",
    "preact/": "https://esm.sh/preact@10.11.0/",
    "preact-render-to-string": "https://esm.sh/*preact-render-to-string@5.2.4",
    "@preact/signals": "https://esm.sh/*@preact/signals@1.0.3",
    "@preact/signals-core": "https://esm.sh/*@preact/signals-core@1.0.1",
    "twind": "https://esm.sh/twind@0.16.17",
    "twind/": "https://esm.sh/twind@0.16.17/",
    "@mantine/core": "https://esm.sh/@mantine/core@5.7.1?alias=react:preact/compat&deps=react",
    "@mantine/hooks": "https://esm.sh/@mantine/hooks@5.7.1?alias=react:preact/compat&deps=react"
  }
}

```

And when I import a Mantine component on a page (`import { TextInput } from '@mantine/core';`), then it even sees the types for the components. But when I try to pop one in, I get this error from Fresh:
 `An error occurred during route handling or page rendering. TypeError: Cannot read properties of undefined (reading 'context')`.

**Reply by klerpi** [reply]
Just tried and got the same error on the newest version

**Reply by swiftsword94** [reply]
I tried it too. not sure but i think the import is pulling something using cjs when there's an esm library available.

**Reply by alexnguyennz** [reply]
For Mantine, using these imports got me past the `reading context` error with `preact@10.11.0`:

```
"@mantine/core": "https://esm.sh/@mantine/core@5.7.1?alias=react:preact/compat&deps=preact@10.11.0",
"@mantine/hooks": "https://esm.sh/@mantine/hooks@5.7.1?alias=react:preact/compat&deps=preact@10.11.0"
```

But components don't work or are unstyled and Tailwind gets messed up.
